### ISSUE TITLE
Facilitate MaaS monitoring for optional services

### ISSUE THREAD

**Issue by cfarquhar** [rcbops/rpc-openstack#1361]
##### On which version(s) do you encounter this issue?

master
###### Please provide the output to 'git describe --all' for  all deployments where you have confirmed it occurs.

```
# git describe --all
heads/master
```
##### What type of issue is this? Please tick the appropriate box.
- [ ] bug - existing functionality that does not work for you as documented
- [x] feature - new functionality that you would like to see added
##### What is the expected behavior?

The rpc_maas role should support configuring checks and alarms for optional services (e.g. ironic-api) via the `openstack_service_local_checks_list` and  `openstack_service_remote_checks_list` lists.  The checks should only be created if the associated service is deployed.  If the associated service is not deployed, the checks should not be created and the playbooks should not throw an exception.

Here are examples of adding local and remote API checks for the Ironic API: [local check](https://github.com/cfarquhar/rpc-openstack/blob/ironic/rpcd/playbooks/roles/rpc_maas/defaults/main.yml#L377), [remote check](https://github.com/cfarquhar/rpc-openstack/blob/ironic/rpcd/playbooks/roles/rpc_maas/defaults/main.yml#L420).
##### What is actually happening?

The tasks in `ensure_local_checks.yml`, `ensure_remote_checks.yaml`, and `ensure_remote_backup_checks.yml` fail with an "error while evaluating conditional" exception when they attempt to evaluate the length or contents a non-existent group.

For example, adding an Ironic API monitor to rpcd/playbooks/roles/rpc_maas/defaults/main.yml like this:

```
openstack_service_remote_checks_list:
[ snip ]
  - { name: "lb_api_check_ironic", group: "ironic_api" }
[ snip ]
```

Results in errors like this:

```
skipping: [aio-cfarquhar01_heat_engine_container-bbba473b] => (item={'group': 'heat_api_cloudwatch', 'name': 'heat_cw_api_check'})
fatal: [aio-cfarquhar01_memcached_container-290b5662] => error while evaluating conditional: inventory_hostname in groups["ironic_api"]
skipping: [aio-cfarquhar01_heat_engine_container-bbba473b] => (item={'group': 'horizon', 'name': 'horizon_local_check'})
fatal: [aio-cfarquhar01_heat_engine_container-bbba473b] => error while evaluating conditional: inventory_hostname in groups["ironic_api"]
```
##### What were you attempting to do when you encountered this error (give as much detail as possible)?

Add monitoring for optional services that will not be deployed in all scenarios.
##### How can we reproduce this issue?

N/A
##### How can we verify your issue is addressed?

MaaS checks for optional services are created if the service is deployed.  If the associated service is not deployed, the checks should not be created and the playbooks should not throw an exception.
##### If you are using or know of a workaround please describe the necessary steps?

One proposed solution can be seen [here](https://github.com/rcbops/rpc-openstack/pull/1354).
##### What is your evaluation on the importance of this issue, and why?

prio-2?  This does not affect any existing installations, but blocks adding monitoring for Ironic.

---
###### For each pull request, update the branch line to reference it.
- [x] master - https://github.com/rcbops/rpc-openstack/pull/1354

##### Upgrade impact. Consider the impact on moving between major, minor and patch releases.
- [x] This change does not impact upgrades
- [ ] This change addresses a potential upgrade impact
- [ ] The upgrade impact is being tracked by another issue.

The issue tracking the upgrade impact is:
##### Testing impact
- [ ] An issue/pull-request has been create or exists to track the testing gap, and is referenced below
- [ ] This change includes additional testing
- [x] This change is tested by the existing setup

Additional testing links:
##### Release note
- [ ] This change includes a release note.
- [x] This change does not require a release note.
##### Does your change require documentation?
- [ ] Yes, and I submitted an issue to [Docs issues](https://github.com/rackerlabs/docs-rpc/issues)
- [ ] Yes, and I contributed a documentation patch to [docs-rpc](https://github.com/rackerlabs/docs-rpc)
- [x] No documentation required for this change

Additional documentation links (please add links if you have ticked yes):

**Comment by git-harry** [comment]
Ironic isn't currently supported by the product, this feature is currently being tracked in https://github.com/rcbops/u-suk-dev/issues/91.

This issue described doesn't block having monitoring for Ironic in that the work to include Ironic in the product would require that monitoring was updated to support it. I don't think rpc_maas was designed with the idea that the vars could be overridden to add additional monitoring for unknown groups.

I've marked this as a low priority enhancement given that it doesn't impact the current product.

**Comment by cfarquhar** [comment]
Thanks @git-harry.  I've added basic MaaS checks and alarms for the Ironic API (https://github.com/rcbops/rpc-openstack/pull/1344) as part of a professional services engagement, and this issue and PR were intended to unblock that PR (or get input on an alternative approach if you guys don't like this one).

I've updated https://github.com/rcbops/u-suk-dev/issues/91 and will attend the triage meeting today.

**Comment by sigmavirus24** [comment]
Am I misunderstanding, or is this just to make our MaaS check/alarm deployment more extendable? This doesn't actually contribute Ironic checks or alarms, yes?

**Comment by cfarquhar** [comment]
Yes, this issue and PR only make MaaS more extendable.  They are a solution to limitations I encountered when writing the actual check and alarm in https://github.com/rcbops/rpc-openstack/pull/1344.  Once the current limitation is resolved, https://github.com/rcbops/rpc-openstack/pull/1344 will be unblocked and I can re-submit it to gating.

Based on my understanding of the contribution guidelines, I separated the changes to make MaaS more flexible into a separate commit, and thus a separate issue and PR.  If that isn't the correct approach, just let me know and I'll wrap this change back in with https://github.com/rcbops/rpc-openstack/pull/1344.

**Comment by git-harry** [comment]
@cfarquhar the dev team uses issues to track current commitments. Every pull request requires its own associated issue. The issues that we work on are decided based on business priorities, the discussions regarding those priorities, for enhancements such as this, happen during the triage meeting held every Monday. Our ability to commit to new work is based on people available and existing commitments. This issue is currently in the backlog and a low priority for the development team based on our current commitments and future commitments to support Ironic. If this issue is a higher priority, please raise it at the next triage meeting and we can judge it against the other priorities of support. I recognise that by including a pull request, you have already completed some of the work however for that to be merged the dev team need to invest time in reviewing and testing it and that is why we must prioritise this issue in the same way as any issue without associated code.

/cc @b3rnard0

**Comment by alextricity25** [comment]
After reading through this, it seems like the primary issue here is not ironic related, but rather a problem with MaaS when trying to add new services to ``openstack_service_local_checks_list`` and ``openstack_service_remote_checks_list lists``. If that's the case, let's go ahead and push this through since it's already been committed to and could come back to bite us in the future.
@Frank-ZhangXin, This should just be a matter of reviewing the connected PR that @cfarquhar has made: https://github.com/rcbops/rpc-openstack/pull/1354.

**Comment by Frank-ZhangXin** [comment]
I think the title of the bug doesn't quite match to what @cfarquhar really want to do here. And it seems that ironic integrating procedure is still going, so there's probably not too much I can help here except reviewing related ironic PR.

**Comment by alextricity25** [comment]
@Frank-ZhangXin, Please review https://github.com/rcbops/rpc-openstack/pull/1354. That PR is not specific to ironic, but rather how the MaaS checks are loaded(It checks to see if the group exists in inventory before dropping in the check template).
The title of this issue is fine. I think @cfarquhar was just using ironic as an example since that is what he was working on at the moment, but really it can be any service.

**Comment by b3rnard0** [comment]
@alextricity25 Okay, so if this is a generic monitoring issue than I don't have to seek @jamiramontez's input on whether this is still a valid change.

**Comment by Frank-ZhangXin** [comment]
Update: PR passed gate tests

**Comment by alextricity25** [comment]
Thanks @Frank-ZhangXin :D. IMO there is no need to backport this since we shouldn't be adding new checks for optional services in previous releases. So unless someone objects, we can move this to "Done".

**Comment by b3rnard0** [comment]
@alextricity25 @git-harry Do we need backports for this work?

cc: @cfarquhar

**Comment by alextricity25** [comment]
@b3rnard0: No, we don't need backports for this.
